### ISSUE TITLE
Align API with Java Streams - few missing methods

### ISSUE THREAD

**Issue by cescoffier** [eclipse/microprofile-reactive#51]
A few methods (or variants) are missing to map with the Java Stream API. Obviously, not all of them make sense, but some may have benefits. Here is the list of methods from Java Streams that we don't provide currently. I've already eliminated `xStream` APIs.

```
boolean	allMatch​(Predicate<? super T> predicate)	
Returns whether all elements of this stream match the provided predicate.

boolean	anyMatch​(Predicate<? super T> predicate)	
Returns whether any elements of this stream match the provided predicate.

<R> R	collect​(Supplier<R> supplier, BiConsumer<R,? super T> accumulator, BiConsumer<R,R> combiner)	
Performs a mutable reduction operation on the elements of this stream.

long	count()	
Returns the count of elements in this stream.

Stream<T>	distinct()	
Returns a stream consisting of the distinct elements (according to Object.equals(Object)) of this stream.

Optional<T>	findAny()	
Returns an Optional describing some element of the stream, or an empty Optional if the stream is empty.

static <T> Stream<T>	iterate​(T seed, Predicate<? super T> hasNext, UnaryOperator<T> next)	
Returns a sequential ordered Stream produced by iterative application of the given next function to an initial element, conditioned on satisfying the given hasNext predicate.

boolean	noneMatch​(Predicate<? super T> predicate)	
Returns whether no elements of this stream match the provided predicate.
```
Some notes:

* `collect` - this is just a different variant
* `findAny` - not sure of the value as we have `findFirst`
* `iterate` - this is just a different variant

**Comment by jroper** [comment]
* I didn't think the `*Match` methods made sense for reactive streams use cases, given the overhead of streaming across the network only to reduce each element to essentially a boolean, if you were doing a database query for example you would implement this in the query itself so that you just returned the result over the network. Also these operators can be effectively achieved with a combination of `filter` and `findFirst`, whereas in JDK8 streams `findFirst` is not desirable due to the inability to do parallel processing with it.
* `findAny` is the same as `findFirst` when the stream is processed sequentially, which reactive streams always are.
* The variant of `collect` that I didn't include is the one that includes the combiner, since the combiner is only useful in the context of parallel processing where you process multiple parts of a collection in parallel and then combine the results. Reactive streams is always sequential, so there's no use for the combiner.
* Reactive streams are very often infinite, making `count` useless. Even when they're not infinite, given the network overhead that's usually associated with reactive streams (because it's most often source over the network), it wouldn't make sense to use `count` as that would require streaming the entire stream over the network for no purpose, much better to do the count at the source and return it via a single response (eg, a REST request or a database call that does a count).
* `distinct` may be useful, though for infinite streams is dangerous.
* Not sure why I didn't include the `iterate` variant.

**Comment by cescoffier** [comment]
* I agree with the `*Match` methods, I never used them (I discovered them yesterday). 
* `findAny` does not make sense, so we can remove it.
* The `collect` variant can omit the combiner and just get the supplier and accumulator. 
* For `count`, yes, as the user can use a collector for this (`counting`)
* `distinct` is dangerous but can be useful, so probably should be added

I will open a PR adding `distinct` and the `iterate` variant. I may open another one with the `collect` variant (without combiner). 

I also think we need to have a section in the spec indicating why we don't have the other methods, and how the user can workaround.

**Comment by cescoffier** [comment]
Found the reason why we don't have the `iterate` variant - it's been added in Java 9.

**Comment by cescoffier** [comment]
Another set of methods:

```
sorted() Returns a stream consisting of the elements of this stream, sorted according to natural order.
sorted(Comparator<? super T> comparator) Returns a stream consisting of the elements of this stream, sorted according to the provided Comparator.
toArray() Returns an array containing the elements of this stream.
toArray(IntFunction<A[]> generator) Returns an array containing the elements of this stream, using the provided generator function to allocate the returned array, as well as any additional arrays that might be required for a partitioned execution or for resizing.

From BaseStream:
S onClose​(Runnable closeHandler)
void close​()
```

I'm not sure about `toArray` as we provide `toList`
For `sorted`, I'm also unconvinced about their utility. But if anyone has a good use case, feel free to comment.

`close` is `cancel`. So basically we have a naming conflict. `cancel` reuses the terminology from Reactive Streams.  
`onClose` can be useful. Now, the question is whether or not the given _handler_ should be called 1) on completion, 2) only when `close` (`cancel`) is called, 3) even if the streams has been cancelled because of an exception.

**Comment by jroper** [comment]
Sorted doesn't make sense for sequential streams, since all the fastest approaches for sorting it always requires buffering the stream first followed by doing a regular sort, so it's no different from using `toList()` followed by `Collections.sort()`. It does make sense for parallel streams since the run time approaches Θ(*n*) as you increase the number of processors available with a parallel merge sort, that's why the method exists on the JDK8 streams API (and indeed, it's given very special treatment if you look at the underlying implementation, they are very careful with ensuring it performs as well as possible.

In Reactive Streams, there is cancel, complete and error, and these all have different meanings. I think close is actually semantically closer to complete, but it all depends on the use case, if you want to ensure that something runs when the stream completes successfully, then complete is what you want to hook into. If you want to always run something when the stream terminates, regardless of whether the signal came from inside the stream due to a error or complete, or if it came from downstream cancelling the operation, then that's a different thing again.

I do think we should consider some of the following:

* `onComplete` - when a stream terminates gracefully (ie, is not cancelled by downstream, and doesn't terminate with an error)
* `onError` - when a stream terminates with an error (but is not cancelled by downstream)
* `onTerminate` - when a stream terminates, regardless of how it terminates
* `recover` - there could be several variants, one possibility is allow the stream to emit a single last element and then terminate successfully, ie this would take a callback of `Throwable -> T` another is to create a new publisher, ie this would take a callback of `Throwable -> PublisherBuilder<T>`.

**Comment by cescoffier** [comment]
Having `onComplete`, `onError` and `onTerminate` make a lot of sense. It supersedes `close` and these stages have clear semantics. 

`recover` is important, even crucial. I wanted to open another issue about error management, but let's use this thread for now. 

If we consider the variants you mentioned we have.... a naming conflict:

* `recover(Function<Throwable, T> errorHandler)`
* `recover(Function<Throwable, PublisherBuilder<T>> errorHandler)`
* `recover(Function<Throwable, Publisher<T>> errorHandler)` (I've added this one)

RX Java uses the following names:

*  onErrorResumeNext(Function<Throwable, Publisher<T>> errorHandler);
*  onErrorReturn(Function<Throwable, T> errorHandler);

(as well as a few variants such as `onExceptionResumeNext`, and methods returning constant values...)

I tend to prefer the Akka stream names: `recover` and `recoverWith`. We can propose:

* `recover(Function<Throwable, T> errorHandler)`
* `recoverWith(Function<Throwable, PublisherBuilder<T>> errorHandler)`
* `recoverWithPublisher(Function<Throwable, Publisher<T>> errorHandler)`

The 2 first names might be switched. 

WDYT? 

In the meantime, let me open PRs for `onComplete`, `onTerminate` and `onError`.

**Comment by jroper** [comment]
I was going to suggest the Akka streams names.

Another alternative to consider is `exceptionally` since that's what `CompletionStage` uses. Ok, now that that's been considered, we can forget about it.

I don't mind the Rx Java names, it's consistent with `onError`. I'd probably say `onErrorResumeWith` rather than `onErrorResumeNext`. Reactor has `onErrorResume` and `onErrorReturn`, though their `onErrorReturn` doesn't get its return value from a function, but uses a static value. Anyway, I don't have a strong opinion on whether we go with Akka, Reactor or RxJava names here.

**Comment by cescoffier** [comment]
I've started to work on the recovery stage in #67

**Comment by jroper** [comment]
I think this is all addressed now.
